lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,377 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2019, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+from collections import OrderedDict, namedtuple
+import itertools
+from statistics import mean
+
+import matplotlib.pyplot as plt
+import pylab as pl
+
+from devlib.target import KernelVersion
+
+from lisa.tests.base import (TestMetric, Result, ResultBundle, TestBundle,
+                             RTATestBundle, CannotCreateError)
+from lisa.target import Target
+from lisa.utils import ArtifactPath, groupby, memoized
+from lisa.wlgen.rta import RTA, RTATask, Periodic, Ramp
+from lisa.trace import FtraceConf, FtraceCollector, requires_events
+from lisa.analysis.rta import RTAEventsAnalysis
+from lisa.analysis.tasks import TaskState, TasksAnalysis
+from lisa.analysis.load_tracking import LoadTrackingAnalysis
+from lisa.datautils import series_integrate
+
+from lisa.tests.scheduler.load_tracking import LoadTrackingHelpers
+
+class UtilTrackingBase(RTATestBundle, LoadTrackingHelpers):
+    """
+    Base class for shared functionality of utilization tracking tests
+    """
+
+    cpufreq_conf = {"governor": "performance"}
+    """
+    The cpufreq configuration used while the synthetic workload is being run.
+    Items are arguments to :meth:`devlib.cpufreq.use_governor`.
+    """
+
+    @classmethod
+    def _from_target(cls, target:Target, *,
+                     res_dir:ArtifactPath=None,
+                     ftrace_coll:FtraceCollector=None) -> 'UtilTrackingBase':
+        plat_info = target.plat_info
+        rtapp_profile = cls.get_rtapp_profile(plat_info)
+
+        # After a bit of experimenting, it turns out that on some platforms
+        # misprediction of the idle time (which leads to a shallow idle state,
+        # a wakeup and another idle nap) can mess up the duty cycle of the
+        # rt-app task we're running. In our case, a 50% duty cycle, 16ms period
+        # task would always be active for 8ms, but it would sometimes sleep for
+        # only 5 or 6 ms.
+        # This is fine to do this here, as we only care about the proper
+        # behaviour of the signal on running/not-running tasks.
+        with target.disable_idle_states():
+            with target.cpufreq.use_governor(**cls.cpufreq_conf):
```

Comment:
BTW: just notice that this cpufreq context manager does not log (at least not at info level) the governor selected. Should we not better add that logging, similar to what we already do for the freezer context manager?

Reply:
I'm a bit torn about that: Adding this logging in LISA means adding a CM directly in Target, rather than just relying on the devlib module. Adding logging in devlib at info level might not be best as it could pollute the output of other applications using devlib ...